pkg/spoke/auth/auth.go

```diff
@@ -97,15 +117,61 @@ func (v *sarValidator) Validate(ctx context.Context, executor *workapiv1.Manifes
 
 	if !allowed {
 		return &NotAllowedError{
-			Err: fmt.Errorf("not allowed to %s the resource %s %s, name: %s",
-				strings.ToLower(string(action)), resource.Group, resource.Resource, resource.Name),
+			Err: fmt.Errorf("not allowed to %s the resource %s %s, %s %s",
+				strings.ToLower(string(action)), resource.Group, resource.Resource, resource.Namespace, resource.Name),
 			RequeueTime: 60 * time.Second,
 		}
 	}
 
+	switch {
+	case action != ApplyAction:
+		return nil
+	case gvr.Group != "rbac.authorization.k8s.io":
+		return nil
+	case gvr.Resource == "roles", gvr.Resource == "rolebindings",
+		gvr.Resource == "clusterroles", gvr.Resource == "clusterrolebindings":
+		// subjectaccessreview can not permission escalation, use an impersonation request to check again
+		return v.checkEscalation(ctx, sa, gvr, namespace, name, obj)
+	}
+
 	return nil
 }
 
+func (v *sarValidator) checkEscalation(ctx context.Context, sa *workapiv1.ManifestWorkSubjectServiceAccount,
+	gvr schema.GroupVersionResource, namespace, name string, obj *unstructured.Unstructured) error {
+
+	dynamicClient, err := v.newImpersonateClientFunc(v.config, username(sa.Namespace, sa.Name))
+	if err != nil {
+		return err
+	}
+
+	_, err = dynamicClient.Resource(gvr).Namespace(namespace).Create(ctx, obj, metav1.CreateOptions{
+		DryRun: []string{"All"},
+	})
+	if apierrors.IsAlreadyExists(err) {
+		return nil
```

Comment:
If the k8s resource exists and its spec is different from the required one, we may need to update the k8s resource and see if the user has the permission or not. What do you think?

Reply:
I checked even if the k8s resource exists, the API server will return the forbidden error when receiving an escalated request.
And I added 2 integration tests to cover these.